src/uwtools/config/tools.py

```diff
@@ -38,6 +38,47 @@ def compare_configs(
     return cfg_a.compare_config(cfg_b.data)
 
 
+def config_check_depths(input_obj: Union[dict, Config], target_format: str, mode: str) -> None:
+    """
+    Check that the depth of the input config does not exceed the target format's max.
+
+    :param input_obj: The input config.
+    :param target_format: The target config.
+    :param mode: The type of config operation. One of 'dump', 'realize' or 'update'.
+    :raises: UWConfigError on excessive input-config depth.
+    """
+    lookup = {
+        "dump",
+        "realize",
+        "update",
+    }
+    if not mode in lookup:
+        raise log_and_error("Mode '%s' should be one of: %s" % (mode, ", ".join(lookup)))
+    cfgclass = format_to_config(target_format)
+    msg = ""
+
+    if cfgclass.DEPTH and not isinstance(input_obj, dict):
+        if mode == "realize" and input_obj.depth != cfgclass.DEPTH:
+            msg = "Cannot write depth-%s input to type-'%s' config" % (
+                input_obj.depth,
+                cfgclass.depth,
+            )
+        if mode == "update" and input_obj.depth > cfgclass.DEPTH:
+            msg = "Cannot update depth-%s input with type-'%s' values" % (
+                input_obj.depth,
+                cfgclass.depth,
+            )
+    if cfgclass.DEPTH and isinstance(input_obj, dict) and mode == "dump":
+        if depth(input_obj) > cfgclass.DEPTH:
+            msg = "Invalid depth-%s input for type-'%s' config" % (
+                depth(input_obj),
+                cfgclass.depth,
+            )
+    if msg:
+        log.error(msg)
+        raise UWConfigError(msg)
```

Comment:
1. This function seems like it would be useful in many contexts, checking not just input objects, but also output objects, value-supplying objects (like those used in rendering templates), etc. What would you think of e.g. `config_obj` as a name instead of `input_obj`, to be more general?
2. Given that callers must specify a `mode` anyway, and therefore must make a choice when calling this function, how about splitting this function into `config_check_depths_dump()`, `config_check_depths_realize()`, and `config_check_depths_update()`? Then the `mode` argument disappears, the magic strings disappear, the validity check on the string disappears, and I think most of the `if` conditionals (there are seven!) disappear. If there's repeated logic between those three functions, it could be DRYed out via one or more private helper functions.

Reply:
1. Ah, good point! I've already generalized `output` as `target`, no need to stop there
2. Funny story, I asked ChatGPT for suggestions after I wrote it and one suggestion was to split it up. I didn't use any of its ideas, but that might be the solution

Reply:
What happens if `input_obj` is an `NMLConfig` and `mode` is `"dump"`? `NMLConfig` has a `DEPTH`, and it is not a `dict`, so wouldn't the `if` on line 60 match? But, then, `"dump"` is not handled in that block. But if I'm trying to `dump()` and `NMLConfig` whose max depth has been exceeded through user mischief, I'd want a check on that. Or have I missed something?

Please ignore this comment while considering breaking this function up into smaller, mode-specific ones -- I suspect that, if that eliminates some of the conditionals then it may be easier to see if we're missing cases we need to handle.

Reply:
I ~think~ that's handled correctly in the split versions now